### ISSUE TITLE
Exit Code is Zero if Subcommand Killed by Signal

### ISSUE THREAD

**Issue by dsanders11** [tj/commander.js#2021]
If a subcommand child process is killed by a signal (e.g. `SIGKILL`), then commander will exit with exit code zero. I'm seeing this where OOM was killing the subcommand but commander was exiting with zero.

Relevant code looks to be here. The arguments to the `close` event are `(code, signal)` and in the case of killed by signal, `code` will be `null`, so `process.exit` gets called with `null` which leads to exiting with zero. Additionally, should the `exitCallback` case (on line 1036) be using `proc.exitCode` instead of `process.exitCode`? The latter doesn't seem correct.

https://github.com/tj/commander.js/blob/4ef19faac1564743d8c7e3ce89ef8d190e1551b4/lib/command.js#L1032-L1038

Possible fix for the issue could be to bubble up signals, although I'm not sure how to handle the `exitCallback` case then, since AFAIK there's no way to get an exit code for a signal from Node.js (other than mapping it yourself).

```js
proc.on('close', (code, signal) => {
  if (signal) {
    process.kill(process.pid, signal);
  } else {
    process.exit.bind(process)(code);
  }
});
```

Alternatively a fixed exit code of 1 could be used in the case of a signal, which would work for the `exitCallback` case. However, that's not desirable since it will hide the underlying exit code from the subcommand.

Manually mapping signals to exit codes might be the simplest way to handle this for both cases.

**Comment by shadowspawn** [comment]
Commander does listen for some signals here: https://github.com/tj/commander.js/blob/4ef19faac1564743d8c7e3ce89ef8d190e1551b4/lib/command.js#L1021

But when I attempted to reproduce, those never fire! I wonder if the signal listener only works within the current process, and instead Commander should be listening for `close` which gets passed the signal and exit code: https://nodejs.org/dist/latest-v18.x/docs/api/child_process.html#event-close

We do listen for `close` already but pay not attention to the parameters.

**Comment by shadowspawn** [comment]
Apologies @dsanders11, you are quite correct. I didn't understand your initial description on first read and made a mess of experimenting with signals. Now your description makes sense, and good detail, thanks. 

I'll take another look! 😊

**Comment by shadowspawn** [comment]
> Additionally, should the `exitCallback` case (on line 1036) be using `proc.exitCode` instead of `process.exitCode`?

Yes, I agree.

**Comment by shadowspawn** [comment]
> Relevant code looks to be here. The arguments to the `close` event are `(code, signal)` and in the case of killed by signal, `code` will be `null`, so `process.exit` gets called with `null` which leads to exiting with zero. 

Ah right, calling `process.exit` with the actual `event` arguments is simplistic at best and wrong for the signal case.

> Possible fix for the issue could be to bubble up signals, 

I am not sure using same signal in parent is an appropriate response in all cases. (Although it would probably reproduce the  eventual exit status.)

> although I'm not sure how to handle the `exitCallback` case then, since AFAIK there's no way to get an exit code for a signal from Node.js (other than mapping it yourself).

The only portable way I found to map signal names to exit codes is `kill -l` (i.e. shell, not node).

However, my understanding after some research is that the external "exit code" for a signal may be more complicated than what the program can achieve by calling `.exit()` anyway, with treatment depending on the shell. 

> Alternatively a fixed exit code of 1 could be used in the case of a signal, which would work for the `exitCallback` case. 

I think this would be simpler to implement, and simpler to predict behaviour.

> However, that's not desirable since it will hide the underlying exit code from the subcommand.

Morally I am not sure if there is an underlying exit code at this point, and in particular this is explicit in what `node` makes available. We get passed an exit code or a signal, one or the other.

**Comment by shadowspawn** [comment]
The related `'exit'` event description is a bit more explicit about the arguments:

> The `'exit'` event is emitted after the child process ends. If the process exited, `code` is the final exit code of the process, otherwise `null`. If the process terminated due to receipt of a signal, `signal` is the string name of the signal, otherwise `null`. One of the two will always be non-`null`.

https://nodejs.org/dist/latest-v18.x/docs/api/child_process.html#event-exit

**Comment by dsanders11** [comment]
Fair points. With this issue I'm concerned about two things, in order of importance.

1. Ensure commander exits with a non-zero exit code if the subcommand process exits due to a signal - otherwise it is silently failing which can cause chained commands (or CI jobs) to continue when they shouldn't
2. An easy way to debug what caused the subcommand to exit

A fixed exit code will address 1, which is good, but does not address 2. You're right that bubbling up the signal to the parent process may not be appropriate in certain corner cases, but it does address 2 in that you get some indication of what caused the subcommand to exit. A fixed exit code wouldn't be as much of a concern if there was a debug log or something (which I don't believe commander has) which could contain the extra information to address 2, but without one, I don't think a fixed exit code can resolve 2.

Given the existing constraints, there may not be a way to address 2, unfortunately.

**Comment by shadowspawn** [comment]
Thanks @dsanders11

I opened draft #2023. Does not have unit tests yet, and I'll leave it draft a couple of days in case we have any fresh ideas about signal number.

**Comment by shadowspawn** [comment]
Released in Commander v12.0.0.
